File: src/main/java/chat/teco/tecochat/chat/application/chat/ChatService.java
package chat.teco.tecochat.chat.application.chat;

import chat.teco.tecochat.chat.application.chat.dto.CopyCommand;
import chat.teco.tecochat.chat.application.chat.dto.UpdateChatTitleCommand;
import chat.teco.tecochat.chat.domain.chat.Chat;
import chat.teco.tecochat.chat.domain.chat.ChatRepository;
import chat.teco.tecochat.chat.domain.chat.event.ChatCopiedEvent;
import lombok.RequiredArgsConstructor;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Transactional
@Service
public class ChatService {

    private final ChatRepository chatRepository;
    private final ApplicationEventPublisher publisher;

    public void updateTitle(UpdateChatTitleCommand command) {
        Chat chat = chatRepository.getById(command.chatId());
        chat.updateTitle(command.memberId(), command.title());
    }

    public Long copy(CopyCommand command) {
        Chat chat = chatRepository.getById(command.chatId());
        Chat copied = chat.copy(command.memberId());
        Long copiedChatId = chatRepository.save(copied).id();
        publisher.publishEvent(new ChatCopiedEvent(chat.id(), copiedChatId));
        return copiedChatId;
    }
}

File: src/main/kotlin/chat/teco/tecochat/application/ChatService.kt
package chat.teco.tecochat.application

import chat.teco.tecochat.domain.chat.ChatCopiedEvent
import chat.teco.tecochat.domain.chat.ChatRepository
import chat.teco.tecochat.domain.chat.getByIdOrThrow
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Transactional
@Service
class ChatService(
    private val chatRepository: ChatRepository,
    private val publisher: ApplicationEventPublisher,
) {

    fun updateTitle(memberId: Long, chatId: Long, request: UpdateChatTitleRequest) {
        val chat = chatRepository.getByIdOrThrow(chatId)
        chat.updateTitle(memberId, request.title)
    }

    fun copy(memberId: Long, chatId: Long): Long {
        val chat = chatRepository.getByIdOrThrow(chatId)
        val copiedChat = chatRepository.save(chat.copy(memberId))
        publisher.publishEvent(ChatCopiedEvent(chat.id, copiedChat.id))
        return copiedChat.id
    }
}

